src/Dimensions.cpp

```diff
@@ -356,24 +386,58 @@ double RangeDimension::tickAt(const ndsize_t index) const {
     return ticks[idx];
 }
 
-
-ndsize_t RangeDimension::indexOf(const double position) const {
-    vector<double> ticks = this->ticks();
+ndsize_t getIndex(const double position, std::vector<double> &ticks, bool lower_bound) {
     if (position < *ticks.begin()) {
         return 0;
     } else if (position > *prev(ticks.end())) {
         return prev(ticks.end()) - ticks.begin();
     }
-    vector<double>::iterator low = std::lower_bound (ticks.begin(), ticks.end(), position);
-    return low - ticks.begin();
+    ndsize_t index;
+    if (lower_bound) {
+        std::vector<double>::iterator lower = std::lower_bound(ticks.begin(), ticks.end(), position);
+        index = lower - ticks.begin();
+    } else{
+        std::vector<double>::iterator upper = std::upper_bound(ticks.begin(), ticks.end(), position) - 1;
+        index = upper - ticks.begin();
+    }
+    return index;
 }
 
 
-vector<double> RangeDimension::axis(const ndsize_t count, const ndsize_t startIndex) const {
+ndsize_t RangeDimension::indexOf(const double position, bool less_or_equal) const {
+    vector<double> ticks = this->ticks();
+    return getIndex(position, ticks, !less_or_equal);
+}
+
+
+pair<ndsize_t, ndsize_t> RangeDimension::indexOf(const double start, const double end) const {
+    vector<double> ticks = this->ticks();
+    ndsize_t si = getIndex(start, ticks, true);
+    ndsize_t ei = getIndex(end, ticks, false);
+    return std::pair<ndsize_t, ndsize_t>(si, ei);
+}
+
+
+std::vector<std::pair<ndsize_t, ndsize_t>> RangeDimension::indexOf(const std::vector<double> &start_positions,
+                                                                   const std::vector<double> &end_positions) const {
+    std::vector<std::pair<ndsize_t, ndsize_t>> indices;
+    vector<double> ticks = this->ticks();
+
+    for (size_t i = 0; i < (std::min(start_positions.size(), end_positions.size())); i++) {
+        if (start_positions[i] > end_positions[i] ) {
+            continue;
```

Comment:
Well, yes, the loop only proceeds as long as we have start and end indices and if start is larger than end, we skip. You are perfectly right, that it would break the match between in- and output. I see only one alternative, i.e. throwing an exception, which would lead to runtime problems I wanted to avoid.

Reply:
Yeah, I agree, throwing an exception wouldn't be nice.
What would happen if it instead returned -1 for each invalid position? Since the return values are indices, a negative index would signal an invalid value.

Not sure I like it entirely but I'd like to hear your thoughts.

Reply:
unfortunately, ndsize_t is unsigned and anything else would be wrong, methinks

Reply:
Well there goes that idea. Yeah, didn't consider that.

Reply:
There is nd**s**size_t in theory, but I am not sure that is a good use case here.

Reply:
na, don't think so. Even though it is unlikely that the full range of ndsize would ever be used in our context, I would rather not use it here. If you think it cleaner, I can change these situations to throw an exception in the cases that 1) the size of the vectors do not match, and 2) if end < start position.
Or, I can leave it as it is :)
In many cases this function is more or internal use, i.e. called via the data access functions of tag and mtag.

Reply:
You make a good point about it being internal only. I'm wondering now how the situation could occur (that end < start). The start and end vectors are built from positions and extents, right?

Reply:
yes, indeed, but since it is part of the public API, there is, at least in principle, the opportunity for errors